File: Python/binary_search.py
import random


def iterative_binary_search(arr, target, low, high):
    while low<=high:
        mid = low + (high - low) // 2

        if target>arr[mid]:
            low = mid + 1
        elif target<arr[mid]:
            high = mid - 1
        else:
            return mid

    return -1


def recursive_binary_search(arr, target, low, high):
    if low<=high:
        mid = low + (high - low) // 2

        if target>arr[mid]:
            return recursive_binary_search(arr, target, mid+1, high)
        elif target<arr[mid]:
            return recursive_binary_search(arr, target, low, mid-1)
        else:
            return mid
    else:
        return -1


if __name__ == '__main__':
    random.seed(0)

    target = 50
    arr = [random.randint(-100,100) for _ in range(10)] + [target]
    arr.sort()

    index_iter = iterative_binary_search(arr, target, 0, len(arr)-1)
    index_rec = recursive_binary_search(arr, target, 0, len(arr)-1)

    print('Iterative search:')
    print('\t Target found at index: ', index_iter)
    print('\t Real index?', target==arr[index_iter])

    print('Recursive search:')
    print('\t Target found at index: ', index_rec)
    print('\t Real index?', target==arr[index_rec])

File: C/binary_search.c
#include <stdio.h>


int IterativeBinarySearch(int arr[], int target, int low, int high){
    while(low<=high){
        int mid = low + (high - low) / 2;

        if(target<arr[mid]){
            high = mid - 1;
        }
        else if(target>arr[mid]){
            low = mid + 1;
        }
        else{
            return mid;
        }
    }

    return -1;
}


int RecursiveBinarySearch(int arr[], int target, int low, int high){
    while(low<=high){
        int mid = low + (high - low) / 2;

        if(target<arr[mid]){
            return RecursiveBinarySearch(arr, target, low, mid-1);
        }
        else if(target>arr[mid]){
            return RecursiveBinarySearch(arr, target, mid+1, high);
        }
        else{
            return mid;
        }
    }

    return -1;
}


void Print(int arr[], int length){
    for(int i=0; i<length; ++i){
        printf("%d ", arr[i]);
    }
    
    printf("\n");
}


int main()
{
    int arr[10] = {-90, -34, -23, -2, 3, 7, 24, 30, 94, 100};
    
    Print(arr, 10);
    int indexIter = IterativeBinarySearch(arr, 30, 0, 10);
    printf("Iterative - Target at index: ");
    printf("%d\n", indexIter);
    
    int indexRec = RecursiveBinarySearch(arr, 30, 0, 10);
    printf("Recursive - Target at index: ");
    printf("%d\n", indexRec);

    return 0;
}
